gobblin-metrics/src/main/java/gobblin/metrics/reporter/ContextAwareReporter.java

```
@@ -0,0 +1,186 @@
+ * multiple calls to start / stop).
+ * </p>
+ */
+@Slf4j
+public class ContextAwareReporter implements Reporter, Closeable {
+
+  private final String name;
+  private boolean started;
+  private final UUID notificationTargetUUID;
+  private final Set<InnerMetricContext> contextsToReport;
+  private final ContextFilter contextFilter;
+
+  public ContextAwareReporter(String name, Config config) {
+    this.name = name;
+    this.started = false;
+    RootMetricContext.get().addNewReporter(this);
+    this.notificationTargetUUID = RootMetricContext.get().addNotificationTarget(new Function<Notification, Void>() {
+      @Nullable @Override public Void apply(Notification input) {
+        notificationCallback(input);
+        return null;
+      }
+    });
+    this.contextFilter = ContextFilterFactory.createContextFilter(config);
+    this.contextsToReport = Sets.newConcurrentHashSet();
+    for(MetricContext context : this.contextFilter.getMatchingContexts()) {
+      this.contextsToReport.add(context.getInnerMetricContext());
+    }
+  }
+
+  /**
+   * Starts the {@link ContextAwareReporter}.
+   * If the {@link ContextAwareReporter} has been started (and not stopped since), this is a no-op.
+   */
+  public final void start() {
+    if(this.started) {
+      log.warn(String.format("Reporter %s has already been started.", this.name));
+      return;
+    }
+    try {
+      startImpl();
+      this.started = true;
+    } catch (Exception exception) {
+      log.warn(String.format("Reporter %s did not start correctly.", this.name), exception);
+    }
+  }
+
+  /**
+   * Actual logic for starting the {@link ContextAwareReporter}. This is a separate method from {@link #start()} to
+   * allow {@link ContextAwareReporter} to handle the started / non-started state.
+   * @throws java.lang.Exception if reporter did not start correctly.
+   */
+  protected void startImpl() throws Exception {
```

This should be either a typed exception or just let them throw RuntimeException. "throws Exception" is meaningless.